keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,489 @@
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
+    // +optional
+    FileRef *FileEnvSource `json:"fileRef,omitempty" protobuf:"bytes,3,opt,name=fileRef"`
+    ...
+}
+
+type FileEnvSource struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // Specify whether the file must exist.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,3,opt,name=optional"`
+}
+```
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file.
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from.  Must be a valid key.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
+### Failure and Fallback Strategy
+
+There are different scenarios in which applying `FileEnvSource` and
+`FileKeySelector` fields may fail. Below are the ones we mapped and their
+outcome once this KEP is implemented.
+
+|Scenario| API Server Result | Kubelet Result |
+|--------|-------------------|----------------|
+|1. The volumeName specified in `FileEnvSource` or `FileKeySelector` field does not exist | Pod creation fails with an error |  |
+|2. The filepath specified in `FileEnvSource` field does not exist | Pod created | Container fails to start and error message in event.|
+|3. The filepath specified in `FileEnvSource` field does not exist but `optional` field is set to true. | Pod created | Container starts and env vars are not populated.|
+|4. Either the filepath or key specified in `FileKeySleector` field does not exist | Pod created | Container fails to start and error message in event.|
+|5. Either the filepath or key specified in `FileKeySleector` field exist but `optional` field is set to true | Pod created | Container starts and env vars are not populated. |
+|6. The specified file is not a parsable env file. | Pod created | Container fails to start and error message in event.|
```

Comment:
I think a new condition needs to be added here. When the specified file contains an illegal environment variable name, we should start the container normally and output the log.

Reply:
Is it different from the number 6 here? and why should we start the container normally? I think the container should fail to start.

Reply:
We have some similar implementations, refer:
https://kubernetes.io/zh-cn/docs/concepts/configuration/secret/#restriction-env-from-invalid
https://github.com/kubernetes/kubernetes/pull/42642

Reply:
Added this failure case.